OpenRA.Mods.RA2/Traits/Mirage.cs

```diff
@@ -0,0 +1,250 @@
+					return self.Owner;
+
+				return self.World.Players.First(p => p.InternalName == mirage.Info.EffectiveOwner);
+			}
+		}
+	}
+
+	[Flags]
+	public enum MirageRevealType
+	{
+		None = 0,
+		Attack = 1,
+		Move = 2,
+		Unload = 4,
+		Infiltrate = 8,
+		Demolish = 16,
+		Damage = 32,
+		Heal = 64,
+		SelfHeal = 128,
+		Dock = 256
+	}
+
+	[Desc("This actor can appear as a different actor in specific situations.")]
+	public class MirageInfo : PausableConditionalTraitInfo
+	{
+		[Desc("Measured in game ticks.")]
+		public readonly int InitialDelay = 10;
+
+		[Desc("Measured in game ticks.")]
+		public readonly int MirageDelay = 30;
+
+		[Desc("Events leading to the actor getting revealed. Possible values are: Attack, Move, Unload, Infiltrate, Demolish, Dock, Damage, Heal and SelfHeal.")]
+		public readonly MirageRevealType RevealOn = MirageRevealType.Attack
+			| MirageRevealType.Unload | MirageRevealType.Infiltrate | MirageRevealType.Demolish | MirageRevealType.Dock;
+
+		[Desc("Map player to use as an effective owner when actor is a mirage.")]
+		public readonly string EffectiveOwner = "Neutral";
+
+		[GrantedConditionReference]
+		[Desc("Granted when the mirage is active.")]
+		public readonly string MirageCondition = null;
+
+		[Desc("Backfall to these actors if none exist in the map.")]
+		[ActorReference]
+		public readonly string[] DefaultTargetTypes = null;
+
+		public override object Create(ActorInitializer init) { return new Mirage(init, this); }
+	}
+
+	public class Mirage : PausableConditionalTrait<MirageInfo>, INotifyDamage, IEffectiveOwner, INotifyUnload, INotifyDemolition, INotifyInfiltration,
+		INotifyAttack, ITick, INotifyCreated, INotifyHarvesterAction
+	{
+		[Sync]
+		private int remainingTime;
+
+		Actor self;
+
+		bool isDocking;
+		ConditionManager conditionManager;
+
+		ActorInfo[] targetTypes;
+
+		CPos? lastPos;
+		bool wasMirage = false;
+		int mirageToken = ConditionManager.InvalidConditionToken;
+
+		public bool Disguised { get { return IsMirage; } }
+
+		public ActorInfo ActorType { get; private set; }
+		public Player Owner { get { return IsMirage ? self.World.Players.First(p => p.InternalName == Info.EffectiveOwner) : null; } }
+
+		public Mirage(ActorInitializer init, MirageInfo info)
+			: base(info)
+		{
+			self = init.Self;
+			remainingTime = info.InitialDelay;
+
+			var targets = self.World.ActorsWithTrait<MirageTarget>().Distinct();
+			targetTypes = targets.Select(a => a.Actor.Info).ToArray();
+
+			if (!targetTypes.Any() && info.DefaultTargetTypes != null)
+				targetTypes = self.World.Map.Rules.Actors.Where(a => info.DefaultTargetTypes.Contains(a.Key)).Select(a => a.Value).ToArray();
+
+			ActorType = targetTypes.RandomOrDefault(self.World.SharedRandom);
+		}
+
+		protected override void Created(Actor self)
+		{
+			conditionManager = self.TraitOrDefault<ConditionManager>();
+
+			if (IsMirage)
+			{
+				wasMirage = true;
+				ActorType = targetTypes.RandomOrDefault(self.World.SharedRandom);
```

Comment:
I don't think we need to do that here anymore after we did in the ctor?

Reply:
For some reason we still have to do this otherwise the sprite vanishes.

Reply:
Actually it is redundant here. The one in the constructor is critical.